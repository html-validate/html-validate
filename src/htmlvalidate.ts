import { Config, ConfigLoader } from './config';
import { Engine, TokenDump, EventDump } from './engine';
import { Parser } from './parser';
import { Report } from './reporter';
import { Source } from './context';

class HtmlValidate {
	private globalConfig: Config;

	constructor(options?: any){
		const defaults = Config.empty();
		this.globalConfig = defaults.merge(options ? Config.fromObject(options) : Config.defaultConfig());
	}

	/**
	 * Parse HTML from string.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateString(str: string): Report {
		const source = {
			data: str,
			filename: 'inline',
			line: 1,
			column: 1,
		};
		return this.validateSource(source);
	}

	/**
	 * Parse HTML from source.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateSource(source: Source): Report {
		const config = this.getConfigFor('inline');
		const engine = new Engine(config, Parser);
		return engine.lint([source]);
	}

	/**
	 * Parse HTML from file.
	 *
	 * @param filename {string} - Filename to read and parse.
	 * @return {object} - Report output.
	 */
	public validateFile(filename: string): Report {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.lint(source);
	}

	public dumpTokens(filename: string): TokenDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTokens(source);
	}

	public dumpEvents(filename: string): EventDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpEvents(source);
	}

	public dumpTree(filename: string): string[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTree(source);
	}

	/**
	 * Get parser for given filename.
	 */
	public getParserFor(source: Source): Parser {
		const config = this.getConfigFor(source.filename);
		return new Parser(config);
	}

	/**
	 * Get configuration for given filename.
	 */
	public getConfigFor(filename: string): Config {
		const loader = new ConfigLoader();
		const config = loader.fromTarget(filename);
		const merged = this.globalConfig.merge(config);
		merged.init();
		return merged;
	}
}

export default HtmlValidate;
