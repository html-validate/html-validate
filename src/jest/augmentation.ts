/** module augmentation: jest.d.ts */
import { type ConfigData } from "../config";
import { type Message } from "../message";

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace -- module augmentation */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/no-empty-object-type, @typescript-eslint/no-unused-vars -- to match jest declaration */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toHaveError(error: Partial<Message>): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(errors: Array<[string, string] | Record<string, unknown>>): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 * @param filename - Optional filename used when matching transformer and
			 * loading configuration.
			 */
			toHTMLValidate(filename?: string): R;
			toHTMLValidate(config: ConfigData, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, config: ConfigData, filename?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with snapshot.
			 */
			toMatchCodeframe(snapshot?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with inline snapshot.
			 */
			toMatchInlineCodeframe(snapshot?: string): R;
		}
	}
}
/** module augmentation end */
