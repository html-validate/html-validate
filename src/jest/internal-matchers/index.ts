export { toBeToken } from "./to-be-token";
