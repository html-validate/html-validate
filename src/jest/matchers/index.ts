export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
