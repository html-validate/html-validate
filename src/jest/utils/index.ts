export { diff } from "./diff";
export { type MaybeAsyncCallback, diverge } from "./diverge";
export { flattenMessages } from "./flatten-messages";
export { isThenable } from "./is-thenable";
export { type MatcherContext } from "./matcher-context";
export { type MatcherExpect } from "./matcher-expect";
export { type MatcherResult } from "./matcher-result";
export { stripAnsi } from "./strip-ansi";
