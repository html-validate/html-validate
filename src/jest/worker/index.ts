export { type ValidateStringFn } from "./worker";
export { createSyncFn } from "./create-sync-fn";
export { default as workerPath } from "./worker.ts?worker&url";
