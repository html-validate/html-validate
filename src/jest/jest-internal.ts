import "./jest";

import { Location } from "../context";
import { TokenType } from "../lexer";
import { toBeToken } from "./internal-matchers";

interface TokenMatcher {
	type: TokenType;
	location?: Partial<Location>;
	data?: any;
}

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-unused-vars */
		interface Matchers<R, T = {}> {
			toBeToken(expected: TokenMatcher): R;
		}
	}
}

expect.extend({
	toBeToken,
});
