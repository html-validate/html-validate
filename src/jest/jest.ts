import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
