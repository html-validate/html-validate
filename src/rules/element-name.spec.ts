import { HtmlValidate } from "../htmlvalidate";
import "../jest";

const DEFAULT_PATTERN = "^[a-z][a-z0-9\\-._]*-[a-z0-9\\-._]*$";

describe("rule element-name", () => {
	let htmlvalidate: HtmlValidate;

	describe("configured with default pattern", () => {
		beforeAll(() => {
			htmlvalidate = new HtmlValidate({
				root: true,
				rules: { "element-name": "error" },
			});
		});

		it("should report error when custom element name does not have a dash", () => {
			expect.assertions(2);
			const report = htmlvalidate.validateString("<foobar></foobar>");
			expect(report).toBeInvalid();
			expect(report).toHaveError("element-name", "<foobar> is not a valid element name");
		});

		it("should report error when custom element name does not start with letter", () => {
			expect.assertions(2);
			const report = htmlvalidate.validateString("<1-foo></1-foo>");
			expect(report).toBeInvalid();
			expect(report).toHaveError("element-name", "<1-foo> is not a valid element name");
		});

		it("should not report error when custom element name is valid", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString("<foo-bar></foo-bar>");
			expect(report).toBeValid();
		});

		it("should not report when using builtin elements", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString("<span><a><span></span></a></span>");
			expect(report).toBeValid();
		});

		it("should not report error for xml namespaces", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString("<xmlns:foo></xmlns:foo>");
			expect(report).toBeValid();
		});

		it("smoketest", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateFile("test-files/rules/element-name.html");
			expect(report).toMatchInlineCodeframe(`
				"error: <foo> is not a valid element name (element-name) at test-files/rules/element-name.html:30:2:
				  28 |
				  29 | <!-- invalid custom names -->
				> 30 | <foo></foo>
				     |  ^^^
				  31 | <1-bar></1-bar>
				  32 |
				Selector: foo
				error: <1-bar> is not a valid element name (element-name) at test-files/rules/element-name.html:31:2:
				  29 | <!-- invalid custom names -->
				  30 | <foo></foo>
				> 31 | <1-bar></1-bar>
				     |  ^^^^^
				  32 |
				Selector: 1-bar"
			`);
		});
	});

	describe("configured with custom pattern", () => {
		beforeAll(() => {
			htmlvalidate = new HtmlValidate({
				root: true,
				rules: { "element-name": ["error", { pattern: "^foo-\\w+$" }] },
			});
		});

		it("should report error when custom element name does not match pattern", () => {
			expect.assertions(2);
			const report = htmlvalidate.validateString("<spam-ham></spam-ham>");
			expect(report).toBeInvalid();
			expect(report).toHaveError("element-name", "<spam-ham> is not a valid element name");
		});

		it("should not report error when custom element name does match pattern", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString("<foo-bar></foo-bar>");
			expect(report).toBeValid();
		});

		it("should not report when using builtin elements", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString("<span><a><span></span></a></span>");
			expect(report).toBeValid();
		});

		it("should not report error for xml namespaces", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateString("<xmlns:foo></xmlns:foo>");
			expect(report).toBeValid();
		});

		it("smoketest", () => {
			expect.assertions(1);
			const report = htmlvalidate.validateFile("test-files/rules/element-name.html");
			expect(report).toMatchInlineCodeframe(`
				"error: <spam-ham> is not a valid element name (element-name) at test-files/rules/element-name.html:27:2:
				  25 | <!-- allowed custom names -->
				  26 | <foo-bar></foo-bar>
				> 27 | <spam-ham></spam-ham>
				     |  ^^^^^^^^
				  28 |
				  29 | <!-- invalid custom names -->
				  30 | <foo></foo>
				Selector: spam-ham
				error: <foo> is not a valid element name (element-name) at test-files/rules/element-name.html:30:2:
				  28 |
				  29 | <!-- invalid custom names -->
				> 30 | <foo></foo>
				     |  ^^^
				  31 | <1-bar></1-bar>
				  32 |
				Selector: foo
				error: <1-bar> is not a valid element name (element-name) at test-files/rules/element-name.html:31:2:
				  29 | <!-- invalid custom names -->
				  30 | <foo></foo>
				> 31 | <1-bar></1-bar>
				     |  ^^^^^
				  32 |
				Selector: 1-bar"
			`);
		});
	});

	it("should ignore whitelisted element", () => {
		expect.assertions(1);
		htmlvalidate = new HtmlValidate({
			rules: { "element-name": ["error", { whitelist: ["foobar"] }] },
		});
		const report = htmlvalidate.validateString("<foobar></foobar>");
		expect(report).toBeValid();
	});

	it("should report error when using blacklisted element", () => {
		expect.assertions(2);
		htmlvalidate = new HtmlValidate({
			root: true,
			rules: { "element-name": ["error", { blacklist: ["foo-bar"] }] },
		});
		const report = htmlvalidate.validateString("<foo-bar></foo-bar>");
		expect(report).toBeInvalid();
		expect(report).toHaveError("element-name", "<foo-bar> element is blacklisted");
	});

	describe("should contain documentation for", () => {
		it("blacklisted element", async () => {
			expect.assertions(1);
			htmlvalidate = new HtmlValidate({
				root: true,
				rules: { "element-name": "error" },
			});
			const context = {
				tagName: "element-name",
				pattern: DEFAULT_PATTERN,
				blacklist: ["element-name"],
			};
			const docs = await htmlvalidate.getRuleDocumentation("element-name", null, context);
			expect(docs).toMatchSnapshot();
		});

		it("element not matching default pattern", async () => {
			expect.assertions(1);
			htmlvalidate = new HtmlValidate({
				root: true,
				rules: { "element-name": "error" },
			});
			const context = {
				tagName: "element-name",
				pattern: DEFAULT_PATTERN,
				blacklist: [] as string[],
			};
			const docs = await htmlvalidate.getRuleDocumentation("element-name", null, context);
			expect(docs).toMatchSnapshot();
		});

		it("element not matching custom pattern", async () => {
			expect.assertions(1);
			htmlvalidate = new HtmlValidate({
				root: true,
				rules: { "element-name": "error" },
			});
			const context = {
				tagName: "element-name",
				pattern: "^foo-.+$",
				blacklist: [] as string[],
			};
			const docs = await htmlvalidate.getRuleDocumentation("element-name", null, context);
			expect(docs).toMatchSnapshot();
		});
	});
});
