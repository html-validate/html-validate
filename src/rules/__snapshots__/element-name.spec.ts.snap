// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule element-name configured with custom pattern smoketest 1`] = `
Array [
  Object {
    "errorCount": 3,
    "filePath": "test-files/rules/element-name.html",
    "messages": Array [
      Object {
        "column": 2,
        "context": Object {
          "blacklist": Array [],
          "pattern": "^foo-\\\\w+$",
          "tagName": "spam-ham",
        },
        "line": 27,
        "message": "<spam-ham> is not a valid element name",
        "offset": 319,
        "ruleId": "element-name",
        "ruleUrl": "https://html-validate.org/rules/element-name.html",
        "selector": "spam-ham",
        "severity": 2,
        "size": 8,
      },
      Object {
        "column": 2,
        "context": Object {
          "blacklist": Array [],
          "pattern": "^foo-\\\\w+$",
          "tagName": "foo",
        },
        "line": 30,
        "message": "<foo> is not a valid element name",
        "offset": 372,
        "ruleId": "element-name",
        "ruleUrl": "https://html-validate.org/rules/element-name.html",
        "selector": "foo",
        "severity": 2,
        "size": 3,
      },
      Object {
        "column": 2,
        "context": Object {
          "blacklist": Array [],
          "pattern": "^foo-\\\\w+$",
          "tagName": "1-bar",
        },
        "line": 31,
        "message": "<1-bar> is not a valid element name",
        "offset": 384,
        "ruleId": "element-name",
        "ruleUrl": "https://html-validate.org/rules/element-name.html",
        "selector": "1-bar",
        "severity": 2,
        "size": 5,
      },
    ],
    "source": "<!-- builtin -->
<div></div>
<table>
	<thead>
		<tr>
			<th>foo</th>
			<th>bar</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td>1</td>
			<td>2</td>
		</tr>
		<tr>
			<td>3</td>
			<td>4</td>
		</tr>
	</tbody>
</table>

<!-- xml namespace -->
<xi:include></xi:include>

<!-- allowed custom names -->
<foo-bar></foo-bar>
<spam-ham></spam-ham>

<!-- invalid custom names -->
<foo></foo>
<1-bar></1-bar>
",
    "warningCount": 0,
  },
]
`;

exports[`rule element-name configured with default pattern smoketest 1`] = `
Array [
  Object {
    "errorCount": 2,
    "filePath": "test-files/rules/element-name.html",
    "messages": Array [
      Object {
        "column": 2,
        "context": Object {
          "blacklist": Array [],
          "pattern": "^[a-z][a-z0-9\\\\-._]*-[a-z0-9\\\\-._]*$",
          "tagName": "foo",
        },
        "line": 30,
        "message": "<foo> is not a valid element name",
        "offset": 372,
        "ruleId": "element-name",
        "ruleUrl": "https://html-validate.org/rules/element-name.html",
        "selector": "foo",
        "severity": 2,
        "size": 3,
      },
      Object {
        "column": 2,
        "context": Object {
          "blacklist": Array [],
          "pattern": "^[a-z][a-z0-9\\\\-._]*-[a-z0-9\\\\-._]*$",
          "tagName": "1-bar",
        },
        "line": 31,
        "message": "<1-bar> is not a valid element name",
        "offset": 384,
        "ruleId": "element-name",
        "ruleUrl": "https://html-validate.org/rules/element-name.html",
        "selector": "1-bar",
        "severity": 2,
        "size": 5,
      },
    ],
    "source": "<!-- builtin -->
<div></div>
<table>
	<thead>
		<tr>
			<th>foo</th>
			<th>bar</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td>1</td>
			<td>2</td>
		</tr>
		<tr>
			<td>3</td>
			<td>4</td>
		</tr>
	</tbody>
</table>

<!-- xml namespace -->
<xi:include></xi:include>

<!-- allowed custom names -->
<foo-bar></foo-bar>
<spam-ham></spam-ham>

<!-- invalid custom names -->
<foo></foo>
<1-bar></1-bar>
",
    "warningCount": 0,
  },
]
`;

exports[`rule element-name should contain contextual documentation for blacklisted element 1`] = `
Object {
  "description": "<element-name> is blacklisted by the project configuration.

The following names are blacklisted:
- element-name",
  "url": "https://html-validate.org/rules/element-name.html",
}
`;

exports[`rule element-name should contain contextual documentation for element not matching custom pattern 1`] = `
Object {
  "description": "<element-name> is not a valid element name. This project is configured to only allow names matching the following regular expression:

- \`^foo-.+$\`",
  "url": "https://html-validate.org/rules/element-name.html",
}
`;

exports[`rule element-name should contain contextual documentation for element not matching default pattern 1`] = `
Object {
  "description": "<element-name> is not a valid element name. If this is a custom element HTML requires the name to follow these rules:

- The name must begin with \`a-z\`
- The name must include a hyphen \`-\`
- It may include alphanumerical characters \`a-z0-9\` or hyphens \`-\`, dots \`.\` or underscores \`_\`.",
  "url": "https://html-validate.org/rules/element-name.html",
}
`;

exports[`rule element-name should contain documentation 1`] = `
Object {
  "description": "This is not a valid element name.",
  "url": "https://html-validate.org/rules/element-name.html",
}
`;
